Refactor Slash Command section (#860)
discordjs/guide: guide/interactions/slash-command-permissions.md

# Slash command permissions

> Need to keep some of your commands safe from prying eyes, or only available to the right people? Slash commands support permission overwrites! For both guild and global commands, you can enable or disable a specific **user** or **role** in a guild from using a command.

Reviewer: ```suggestion
Slash commands have their own permissions system, which allows you to control who has access to use which commands. Unlike the slash commands permission setting within the Discord client, you can fine-tune access to commands without preventing the selected user or role from using all commands.
```

> Need to keep some of your commands safe from prying eyes, or only available to the right people? Slash commands support permission overwrites! For both guild and global commands, you can enable or disable a specific **user** or **role** in a guild from using a command.

Author: > Unlike the slash commands permission setting within the Discord client

which permission setting? is it referring to the defaultPermission? Also, I believe this is outdated because it says "with the Discord client". The old revision of the guide used the main library to register commands with a `!deploy` in `messageCreate` event.

I would suggest:

> Slash commands have their own permissions system, which allows you to control which user or role has access to use specific commands.

> Need to keep some of your commands safe from prying eyes, or only available to the right people? Slash commands support permission overwrites! For both guild and global commands, you can enable or disable a specific **user** or **role** in a guild from using a command.

Author: May I suggest:

> Slash commands have their own permission system, which allows you to control who has access to use which commands. You can enable or disable a specific user or role from using a specific command.

> Need to keep some of your commands safe from prying eyes, or only available to the right people? Slash commands support permission overwrites! For both guild and global commands, you can enable or disable a specific **user** or **role** in a guild from using a command.

Reviewer: Seems good to me

You can set the `default_permission` of a slash command with the `setDefaultPermission()` method of the `SlashCommandBuilder()` builder that you have been using so far. For example, if you want to do that for the `ping` command:

```js:no-line-numbers {4}
new SlashCommandBuilder()
  .setName('ping')
  .setDescription('Replies with pong!')
  .setDefaultPermission(false) // defaults to "true" if not specified
```

Since the `default_permission` will either allow or disallow everyone from using a specific slash command, there are two approaches for controlling the permission of a slash command:

- [Denylist](https://en.wikipedia.org/wiki/Blacklist_(computing))
- [Allowlist](https://en.wikipedia.org/wiki/Whitelisting)

The **denylist** approach ENABLES the commnd for everyone (i.e. `setDefaultPermission(true)`), and then manually DISABLE the command for a specific **user** or **role** in a specific guild.

The **allowlist** approach does the opposite. It DISABLES the command for everyone (i.e. `setDefaultPermission(false)`), and then manually ENABLE the command for a specific **user** or **role** in a specific guild.

> The process of "manually enabling/disabling the command for a specific user/role" is called **permission overwrite**.

Author: How about
```
When you manually enable or disable a command for a specific user or role, you're performing a **permission overwrite**.
```

> The process of "manually enabling/disabling the command for a specific user/role" is called **permission overwrite**.

Reviewer: Sure:
```suggestion
When you enable or disable a command for a specific user or role, you perform a **permission overwrite**.
```

![](./images/disabled-command.png)

Next, let's enable the `ping` command for a specific role. You need three IDs for a permission overwrite:

- Guild ID
- Command ID
- User/role ID

So far, you already have the Guild ID in the `deploy-commands.js` file.

- Guild ID ✅
- Command ID
- User/role ID

With these three IDs, you make another HTTP PUT request but to a different endpoint (route). As you might've guessed, that means we need to go back to your `deploy-commands.js` file to implement a few more lines of code:

:::: code-group
::: code-group-item deploy-commands.js
<!-- eslint-skip -->
```js {7,16,23,25-36,38-41,43}
const { SlashCommandBuilder } = require('@discordjs/builders');
const { REST } = require('@discordjs/rest');
const { Routes } = require('discord-api-types/v9');
const { clientId, guildId, token } = require('./config.json');

const commands = [
	new SlashCommandBuilder().setName('ping').setDescription('Replies with pong!').setDefaultPermission(false),
	new SlashCommandBuilder().setName('server').setDescription('Replies with server info!'),
	new SlashCommandBuilder().setName('user').setDescription('Replies with user info!'),

> ].map(command => command.toJSON());

Author: Just to make sure I understood what you meant here: you're suggesting replacing the entire array with a `// ...`?

Would you mind elaborating on the reason behind this suggestion?

> ].map(command => command.toJSON());

Reviewer: The guide, in order to reduce redundancy, uses `// ...` whenever a chunk of code is the same as earlier. Moreover, This `commands` constant can differ based on the user and what they have in their file so it's better to use this, imo.

Lmk what you think.

> ].map(command => command.toJSON());

Author: While i agree with reducing redundancy with something like `// ...`, i ultimately decided to utilize line highlighting to show the relevant piece of code. I know the guide expects the reader to have basic knowledge of programming and JS, but that's not the case in reality (unfortunately). As a result, I believe showing the complete code block is more "beginner friendly".

I would love to refactor all the code blocks i've written to use something like `// ...`, but i'm afraid this might cause confusion for majority of people reading the guide.

If we assume that the reader is able to "fill in the gap" when looking at something like `// ...`, then I think it's also safe to assume that the reader can replace `const commands = [ new SlashCommandBuilder()... ]` with their own code.

> const commands = [
> 	new SlashCommandBuilder().setName('ping').setDescription('Replies with pong!').setDefaultPermission(false),
> 	new SlashCommandBuilder().setName('server').setDescription('Replies with server info!'),
> 	new SlashCommandBuilder().setName('user').setDescription('Replies with user info!'),
> ].map(command => command.toJSON());

Reviewer: Hmm, true.

Maybe:
```suggestion
const commands = [
  // ...
];
```

> ].map(command => command.toJSON());

Author: Alright. So, I guess I should also remove `const { SlashCommandBuilder } = require('@discordjs/builders');`, correct?

const rest = new REST({ version: '9' }).setToken(token);

(async () => {
	try {
		const response = await rest.put(
			Routes.applicationGuildCommands(clientId, guildId),
			{ body: commands },
		);

		console.log('Successfully registered application commands.');

		const pingCommand = response.find(cmd => cmd.name === 'ping');

		const fullPermissions = [
			{
				id: pingCommand.id,
				permissions: [
					{
						id: '1234567890',
						type: 1,
						permission: true,
					},
				],
			},
		];

		await rest.put(
			Routes.guildApplicationCommandsPermissions(clientId, guildId),
			{ body: fullPermissions },
		);

		console.log('Successfully applied permission overwrite.');
	} catch (error) {
		console.error(error);
	}
})();
```
:::
::::

::: warning
We are not using the command handler version of the `deploy-commands.js` file for the sake of simplicity during explanation. However, all the concepts and examples in this page are also applicable for the command handler version of the aformentioned file.
:::

> As you can see in line 16, calling `rest.put()` to this specific endpoint `Routes.applicationGuildCommands()` will return a response. We store that response in the variable called `response`. This response will be an array of all the registered commands.
> 

Reviewer: ```suggestion
Here, we are taking the response of the first `PUT` method which contains an array of all the registered commands.
```

> Then, in line 23, we retrieve the `ping` command from that array with [`Array.find()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find). Once you retrieve the `ping` command, you can access its ID from the `id` property.

Reviewer: ```suggestion
Then, we retrieve the `ping` command from that array using the [`Array.find()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) method. Once you retrieve the `ping` command, you can access its id from the `id` property.
```

> - Guild ID ✅
> - Command ID ✅
> - User/role ID
> 

Reviewer: ```suggestion

```

> Next, in line 25 to 33, you define an array of JSON objects. This JSON needs to have two properties: `id` and `permissions`. The `id` should be the command ID (line 27) and the `permissions` should be an array of [Application Command Permissions Structure](https://discord.com/developers/docs/interactions/application-commands#application-command-permissions-object-application-command-permissions-structure) objects (line 28-34). 

Reviewer: ```suggestion
Next, we define an array of objects. Each object needs to have two properties: `id` and `permissions`. The `id` should be the command's id and the `permissions` should be an array of [Application Command Permissions Structure](https://discord.com/developers/docs/interactions/application-commands#application-command-permissions-object-application-command-permissions-structure) objects. 
```

> This `Application Command Permissions Structure` object needs to have three properties: `id`, `type`, and `permission`. The `id` will be the user id or role id that you want to allowlist (line 30). The `type` specifies what type of `id` it is (line 31). Type `1` means `ROLE` id, and type `2` means `USER` id. Finally, the `permission` specifies if you want to allow or disallow this `id` to use the command (line 32). If `permission` is `true` then it means allow, if it is `false` then it means disallow. Since your `ping` command is already registered with `setDefaultPermission(false)`, it wouldn't make much sense to also have a permission overwrite to disallow specific users or roles. On top of that, you are using the allowlist approach, which means you are enabling the command for a specific role, hence you should set `permission: true`.

Reviewer: ```suggestion
This `Application Command Permissions Structure` object needs to have three properties: `id`, `type`, and `permission`. The `id` will be the user id or role id that you want to allow running the command. The `type` specifies what type of `id` it is. Type `1` is for roles, and type `2` is for users. Finally, the `permission` specifies if you want to allow or disallow this `id` to use the command. If `permission` is `true` then it means allow, if it is `false` then it means disallow. Since we want to allow users, we set `permission` to true.
```

> - Guild ID ✅
> - Command ID ✅
> - User/role ID ✅
> 

Reviewer: ```suggestion

```

> Finally, you make the HTTP PUT request in line 38. However, you need a different endpoint (route) this time: `Routes.guildApplicationCommandsPermissions()` (line 39), and you pass your `fullPermissions` array in the body of the PUT request (line 40).

Reviewer: ```suggestion
Finally, we make a `PUT` request to update command overwrites. We use the `Routes.guildApplicationCommandsPermissions()` route, and pass our `fullPermissions` array as the body of the `PUT` request.
```

> You can go ahead and run the deployment script again:

Reviewer: ```suggestion
We can now run the deployment script again:
```

```sh:no-line-numbers
node deploy-commands.js
```

> Congratulations! 🎉
> 

Reviewer: ```suggestion

```

> Your permission overwrite has been applied. You can go ahead and assign yourself that role in your guild. You should be able to use the `ping` command now. If you remove that role from yourself, you will see the `ping` command as disabled again.

Reviewer: ```suggestion
Congratulations, you've successfully applied your first permission overwrites. You can go ahead and assign yourself the role you allowed in your guild. You should be able to use the `ping` command now. If you remove that role from yourself, you should see the `ping` command disabled again.
```

::: warning

> As you've probably noticed, you can only adjust the permissions of a command at the top-level. This means that you do not have the granularity of adjusting permissions for subcommand groups and/or subcommands.

Reviewer: ```suggestion
You can only adjust the permissions of commands at the top level. This means that you can not adjust permissions for subcommand groups and/or subcommands.
```

> Let's use the `/permissions` command from [Subcommands and Subcommand groups](#subcommands-and-subcommand-groups) as example. It's NOT possible to enable `/permissions user get` for a user, while also disabling `/permissions user edit` for that same user.

Reviewer: ```suggestion
Let's use the `/permissions` command from the [Subcommands and Subcommand groups](#subcommands-and-subcommand-groups) as an example. It is not possible to enable `/permissions user get` for a user, while also disabling `/permissions user edit` for that same user.
```

:::

If you want to disallow a user from using your `/server` slash command, simply implement the following:

```js:no-line-numbers {2,15-24}
const pingCommand = response.find(cmd => cmd.name === 'ping');
const serverCommand = response.find(cmd => cmd.name === 'server');

const fullPermissions = [
	{
		id: pingCommand.id,
		permissions: [
			{
				id: '1234567890',
				type: 1,
				permission: true,
			},
		],
	},
	{
		id: serverCommand.id,
		permissions: [
			{
				id: '1234567890',
				type: 2,
				permission: false,
			},
		],
	},
];
```

### Alternative

You can also make permission overwrite with the bot itself.

```js:no-line-numbers {5}
const pingCommand = client.guilds.cache.get(guildId).commands.cache.find(cmd => cmd.name === 'ping');

const permissions: [
  {
    id: '1234567890',
    type: 'ROLE',
    permission: true,
  }
];

await pingCommand.permissions.set({ permissions });
```

> All three IDs are satisfied here as well:
> 
> - Guild ID ✅
> - Command ID ✅
> - User/role ID ✅
> 
> The Guild ID is specified in `guildId`.
> 
> The Command ID is built into the `pingCommand` variable, because it's an instance of [ApplicationCommand](https://discord.js.org/#/docs/main/stable/class/ApplicationCommand).
> 
> The user/role ID is specified in the highlighted line.

Reviewer: ```suggestion
And that's all you need to know about slash command permissions!
```
